File: old-project/backend/models/notification.go
package models

import (
	"errors"
	"fmt"
	"time"

	"gorm.io/gorm"
)

// Notification represents an in-app notification
type Notification struct {
	gorm.Model
	RecipientID uint   `json:"recipient_id" gorm:"not null"`
	RecipientRole string `json:"recipient_role" gorm:"type:TEXT CHECK(recipient_role IN ('student','professor'));not null"`
	Title       string `json:"title" gorm:"not null"`
	Message     string `json:"message" gorm:"type:text;not null"`
	Type        string `json:"type" gorm:"type:TEXT CHECK(type IN ('info','success','warning','error'));default:'info'"`
	Read        bool   `json:"read" gorm:"default:false"`
	ReadAt      *time.Time `json:"read_at"`
}

// CreateNotification creates a new notification
func CreateNotification(db *gorm.DB, recipientID uint, recipientRole, title, message, notifType string) (*Notification, error) {
	// Validate recipient role
	if recipientRole != "student" && recipientRole != "professor" {
		return nil, errors.New("invalid recipient role")
	}

	// Validate type
	if notifType == "" {
		notifType = "info"
	}
	if notifType != "info" && notifType != "success" && notifType != "warning" && notifType != "error" {
		return nil, errors.New("invalid notification type")
	}

	notification := &Notification{
		RecipientID:   recipientID,
		RecipientRole: recipientRole,
		Title:         title,
		Message:       message,
		Type:          notifType,
		Read:          false,
	}

	if err := db.Create(notification).Error; err != nil {
		return nil, err
	}

	return notification, nil
}

// GetNotificationsByRecipient returns all notifications for a specific user
func GetNotificationsByRecipient(db *gorm.DB, recipientID uint, recipientRole string, unreadOnly bool) ([]Notification, error) {
	var notifications []Notification
	query := db.Where("recipient_id = ? AND recipient_role = ?", recipientID, recipientRole)
	
	if unreadOnly {
		query = query.Where("read = ?", false)
	}
	
	err := query.Order("created_at DESC").Find(&notifications).Error
	return notifications, err
}

// MarkNotificationAsRead marks a notification as read
func MarkNotificationAsRead(db *gorm.DB, notificationID, recipientID uint) error {
	now := time.Now()
	result := db.Model(&Notification{}).
		Where("id = ? AND recipient_id = ?", notificationID, recipientID).
		Updates(map[string]interface{}{
			"read":    true,
			"read_at": &now,
		})

	if result.Error != nil {
		return result.Error
	}
	if result.RowsAffected == 0 {
		return errors.New("notification not found or unauthorized")
	}
	return nil
}

// MarkAllNotificationsAsRead marks all notifications as read for a user
func MarkAllNotificationsAsRead(db *gorm.DB, recipientID uint, recipientRole string) error {
	now := time.Now()
	return db.Model(&Notification{}).
		Where("recipient_id = ? AND recipient_role = ? AND read = ?", recipientID, recipientRole, false).
		Updates(map[string]interface{}{
			"read":    true,
			"read_at": &now,
		}).Error
}

// DeleteNotification deletes a notification
func DeleteNotification(db *gorm.DB, notificationID, recipientID uint) error {
	result := db.Where("id = ? AND recipient_id = ?", notificationID, recipientID).Delete(&Notification{})
	if result.Error != nil {
		return result.Error
	}
	if result.RowsAffected == 0 {
		return errors.New("notification not found or unauthorized")
	}
	return nil
}

// GetUnreadCount returns the count of unread notifications
func GetUnreadNotificationCount(db *gorm.DB, recipientID uint, recipientRole string) (int64, error) {
	var count int64
	err := db.Model(&Notification{}).
		Where("recipient_id = ? AND recipient_role = ? AND read = ?", recipientID, recipientRole, false).
		Count(&count).
		Error
	return count, err
}

// NotifyApplicationStatusChange notifies a student when their application status changes
func NotifyApplicationStatusChange(db *gorm.DB, studentID uint, opportunityName, status string) error {
	title := "Application Update"
	message := fmt.Sprintf("Your application for '%s' has been %s", opportunityName, status)
	notifType := "info"
	
	if status == "accepted" {
		title = "🎉 Application Accepted!"
		notifType = "success"
	} else if status == "rejected" {
		message = fmt.Sprintf("Your application for '%s' was not accepted this time", opportunityName)
	}
	
	_, err := CreateNotification(db, studentID, "student", title, message, notifType)
	return err
}

// NotifyNewReport notifies a professor when a student submits a weekly report
func NotifyNewReport(db *gorm.DB, professorID uint, studentName string) error {
	title := "📝 Weekly Report Submitted"
	message := fmt.Sprintf("%s submitted a weekly report", studentName)
	_, err := CreateNotification(db, professorID, "professor", title, message, "info")
	return err
}

File: lib/services/notification.ts
import prisma from "@/lib/prisma";

export type RecipientRole = "student" | "professor";
export type NotificationType = "info" | "success" | "warning" | "error";

/**
 * Create a notification
 */
export async function createNotification(
  recipientId: bigint,
  recipientRole: RecipientRole,
  title: string,
  message: string,
  type: NotificationType = "info"
) {
  // Validate recipient role
  if (!["student", "professor"].includes(recipientRole)) {
    throw new Error("Invalid recipient role");
  }

  // Validate type
  if (!["info", "success", "warning", "error"].includes(type)) {
    throw new Error("Invalid notification type");
  }

  const notification = await prisma.notification.create({
    data: {
      recipientId,
      recipientRole,
      title,
      message,
      type,
      read: false,
    },
  });

  return notification;
}

/**
 * Get notifications by recipient
 */
export async function getNotificationsByRecipient(
  recipientId: bigint,
  recipientRole: string,
  unreadOnly = false
) {
  return prisma.notification.findMany({
    where: {
      recipientId,
      recipientRole,
      ...(unreadOnly ? { read: false } : {}),
    },
    orderBy: { createdAt: "desc" },
  });
}

/**
 * Mark a notification as read
 */
export async function markNotificationAsRead(
  notificationId: bigint,
  recipientId: bigint
) {
  const result = await prisma.notification.updateMany({
    where: {
      id: notificationId,
      recipientId,
    },
    data: {
      read: true,
      readAt: new Date(),
    },
  });

  if (result.count === 0) {
    throw new Error("Notification not found or unauthorized");
  }
}

/**
 * Mark all notifications as read for a user
 */
export async function markAllNotificationsAsRead(
  recipientId: bigint,
  recipientRole: string
) {
  await prisma.notification.updateMany({
    where: {
      recipientId,
      recipientRole,
      read: false,
    },
    data: {
      read: true,
      readAt: new Date(),
    },
  });
}

/**
 * Delete a notification
 */
export async function deleteNotification(
  notificationId: bigint,
  recipientId: bigint
) {
  const result = await prisma.notification.deleteMany({
    where: {
      id: notificationId,
      recipientId,
    },
  });

  if (result.count === 0) {
    throw new Error("Notification not found or unauthorized");
  }
}

/**
 * Get unread notification count
 */
export async function getUnreadNotificationCount(
  recipientId: bigint,
  recipientRole: string
) {
  return prisma.notification.count({
    where: {
      recipientId,
      recipientRole,
      read: false,
    },
  });
}

/**
 * Notify professor when a student submits a weekly report
 */
export async function notifyNewReport(
  professorId: bigint,
  studentName: string
) {
  const title = "📝 Weekly Report Submitted";
  const message = `${studentName} submitted a weekly report`;
  return createNotification(professorId, "professor", title, message, "info");
}
